### PR TITLE
Pass parent to dataflow node constructor. 

### LINKED ISSUE #2197: Why isn't each type of Node constructor include parent?
It seems like we always know its parent when we build things.

Including parent in the constructor will make the code shorter and further avoid typo bugs.

### PR DESCRIPTION
**Pull request by domoritz**
Fixes #2197.

### CODE REVIEW

File: src/compile/data/parse.ts [hunk 1]
```diff
@@ -179,129 +157,88 @@ Description of the dataflow (http://asciiflow.com/):
...
-    for (const geojson of GeoJSONNode.makeAll(model)) {
-      geojson.parent = head;
-      head = geojson;
-    }
-
-    for (const geopoint of GeoPointNode.makeAll(model)) {
-      geopoint.parent = head;
-      head = geopoint;
-    }
+    head = GeoJSONNode.makeAll(head, model);
```
Comment (kanitw) [inline src/compile/data/parse.ts]: makeAll gives a feel like factory pattern, but I think it's better named `parseAll()` ? (More consistent with `parseTransformArray` above.)

File: src/compile/data/bin.ts [hunk 2]
```diff
@@ -81,14 +81,14 @@ export interface BinComponent {
...
   }
 
-  constructor(private bins: Dict<BinComponent>) {
-    super();
+  constructor(parent: DataFlowNode, private bins: Dict<BinComponent>) {
+    super(parent);
   }
 
-  public static makeBinFromEncoding(model: ModelWithField) {
+  public static makeBinFromEncoding(parent: DataFlowNode, model: ModelWithField) {
```
Comment (kanitw) [inline src/compile/data/bin.ts]: Rename to makeFromEncoding?

File: src/compile/data/parse.ts [hunk 3]
```diff
@@ -179,129 +157,88 @@ Description of the dataflow (http://asciiflow.com/):
...
 
   if (model.transforms.length > 0) {
-    const {first, last} = parseTransformArray(model);
-    first.parent = head;
-    head = last;
+    head = parseTransformArray(head, model);
   }
 
-  const parse = ParseNode.make(model);
+  const parse = ParseNode.make(head, model);
```
Comment (kanitw) [inline src/compile/data/parse.ts]:
```
head = ParseNode.make(head, model) || head;  
```

?

File: src/compile/data/stack.ts [hunk 4]
```diff
@@ -56,16 +56,16 @@ export class StackNode extends DataFlowNode {
...
 
   public clone() {
-    return new StackNode(duplicate(this._stack));
+    return new StackNode(null, duplicate(this._stack));
   }
 
-  constructor(stack: StackComponent) {
-    super();
+  constructor(parent: DataFlowNode, stack: StackComponent) {
+    super(parent, );
```
Comment (kanitw) [inline src/compile/data/stack.ts]: remove comma

File: src/compile/data/parse.ts (hunk 1 above)
Comment (domoritz) [reply src/compile/data/parse.ts]: Done

File: src/compile/data/bin.ts (hunk 2 above)
Comment (domoritz) [reply src/compile/data/bin.ts]: Done

File: src/compile/data/parse.ts (hunk 3 above)
Comment (domoritz) [reply src/compile/data/parse.ts]: Nope, we need parse below.

File: src/compile/data/stack.ts (hunk 4 above)
Comment (domoritz) [reply src/compile/data/stack.ts]: Done

**Pull request merged**